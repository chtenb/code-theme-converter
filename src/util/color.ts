const toNumber = (r: number) => (v: string) => parseInt(v, r)

const match = (m: RegExp) => (str: string) => str.match(m)

const splitChannels = match(/.{2}/g)

const fromHex = toNumber(16)

const toPercentage = (v: string): number => ((fromHex(v) / 255) * 1000) / 1000

export function hex2Rgb (hex: string): number[] {
  return (
    hex
      .replace(
        /^#?([a-f\d])([a-f\d])([a-f\d])$/i,
        (_m, r: string, g: string, b: string) => '#' + r + r + g + g + b + b
      )
      .substring(1)
      .match(/.{2}/g)
      ?.map(x => parseInt(x, 16)) ?? [0, 0, 0]
  )
}

export function hex2Rgba (hex: string): number[] {
  const num = parseInt(hex.slice(1), 16)
  return [(num >> 16) & 255, (num >> 8) & 255, num & 255, (num >> 24) & 255]
}

export function getAlphaFromHex (hex: string): number | undefined {
  return hex2Rgba(hex).pop()
}

export function removeHashFromHex (hex: string): string {
  return hex.includes('#') ? hex.substr(1) : hex
}

export function eightDigitHex2Rgba (hex: string): number[] {
  const hashless = removeHashFromHex(hex)
  const c = splitChannels(hashless) ?? []
  const alpha = c[3] != null ? toPercentage(c[3]) : 1.0
  return [...c.slice(0, 3).map(fromHex), alpha]
}
