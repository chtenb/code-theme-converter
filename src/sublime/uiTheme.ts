import { CodeTheme, findEditorColor } from '../util/vscode'
import {
  SublimeUIClass,
  SublimeUISettings,
  SublimeUITheme,
  SublimeUIAttributes
} from '../types/sublime'

export function createAdaptiveTheme ({ colors }: CodeTheme): SublimeUITheme {
  const findEditorColorForField = findEditorColor(colors)

  return {
    extends: 'Adaptive.sublime-theme',
    variables: {
      font_face: SublimeUISettings.SYSTEM,
      font_size_sm: 11,
      font_size: 12,
      font_size_lg: 13
    },
    rules: [
      {
        class: SublimeUIClass.title_bar,
        fg: findEditorColorForField(['titleBar.activeForeground'])
      },
      {
        class: SublimeUIClass.sidebar_container,
        'layer0.tint': findEditorColorForField(['sideBar.background'])
      },
      {
        class: SublimeUIClass.status_bar,
        'layer0.tint': findEditorColorForField(['statusBar.background'])
      },
      {
        class: SublimeUIClass.label_control,
        parents: [{ class: SublimeUIClass.status_bar }],
        fg: findEditorColorForField(['statusBar.foreground'])
      },
      {
        class: SublimeUIClass.tree_row,
        attributes: [SublimeUIAttributes.selectable, SublimeUIAttributes.hover],
        'layer0.tint': findEditorColorForField([
          'list.inactiveSelectionBackground'
        ]),
        'layer0.opacity': 1.0
      },
      {
        class: SublimeUIClass.tree_row,
        attributes: [SublimeUIAttributes.selected],
        'layer0.tint': findEditorColorForField([
          'list.activeSelectionBackground'
        ]),
        'layer0.opacity': 1.0
      },
      {
        class: SublimeUIClass.sidebar_heading,
        fg: findEditorColorForField(['sideBarSectionHeader.foreground'])
      },
      {
        class: SublimeUIClass.sidebar_label,
        fg: findEditorColorForField(['sideBar.foreground'])
      },
      {
        class: SublimeUIClass.sidebar_label,
        parents: [
          {
            class: SublimeUIClass.tree_row,
            attributes: [SublimeUIAttributes.selected]
          }
        ],
        fg: findEditorColorForField(['list.activeSelectionForeground'])
      },
      {
        class: SublimeUIClass.vcs_status_badge,
        parents: [
          {
            class: SublimeUIClass.file_system_entry,
            attributes: [SublimeUIAttributes.untracked]
          }
        ],
        'layer0.tint': findEditorColorForField([
          'gitDecoration.untrackedResourceForeground'
        ])
      },
      {
        class: SublimeUIClass.vcs_status_badge,
        parents: [
          {
            class: SublimeUIClass.file_system_entry,
            attributes: [SublimeUIAttributes.modified]
          }
        ],
        'layer0.tint': findEditorColorForField([
          'gitDecoration.modifiedResourceForeground'
        ])
      },
      {
        class: SublimeUIClass.vcs_status_badge,
        parents: [
          {
            class: SublimeUIClass.file_system_entry,
            attributes: [SublimeUIAttributes.missing]
          }
        ],
        'layer0.tint': findEditorColorForField([
          'gitDecoration.deletedResourceForeground'
        ])
      },
      {
        class: SublimeUIClass.vcs_status_badge,
        parents: [
          {
            class: SublimeUIClass.file_system_entry,
            attributes: [SublimeUIAttributes.staged]
          }
        ],
        'layer0.tint': findEditorColorForField([
          'gitDecoration.addedResourceForeground'
        ])
      },
      {
        class: SublimeUIClass.vcs_status_badge,
        parents: [
          {
            class: SublimeUIClass.file_system_entry,
            attributes: [SublimeUIAttributes.added]
          }
        ],
        'layer0.tint': findEditorColorForField([
          'gitDecoration.addedResourceForeground'
        ])
      },
      {
        class: SublimeUIClass.vcs_status_badge,
        parents: [
          {
            class: SublimeUIClass.file_system_entry,
            attributes: [SublimeUIAttributes.deleted]
          }
        ],
        'layer0.tint': findEditorColorForField([
          'gitDecoration.deletedResourceForeground'
        ])
      },
      {
        class: SublimeUIClass.sidebar_label,
        parents: [
          {
            class: SublimeUIClass.file_system_entry,
            attributes: [SublimeUIAttributes.ignored]
          }
        ],
        color: findEditorColorForField([
          'gitDecoration.ignoredResourceForeground'
        ])
      },
      {
        class: SublimeUIClass.tab_control,
        attributes: [SublimeUIAttributes.selected],
        tint_modifier: findEditorColorForField(['tab.activeBackground'])
      },
      {
        class: SublimeUIClass.tab_control,
        attributes: [SublimeUIAttributes.dirty],
        settings: [SublimeUISettings.highlight_modified_tabs],
        'layer2.tint': findEditorColorForField(['list.highlightForeground'])
      },
      {
        class: SublimeUIClass.tab_label,
        fg: findEditorColorForField(['sideBar.foreground'])
      },
      {
        class: SublimeUIClass.tab_label,
        parents: [
          {
            class: SublimeUIClass.tab_control,
            attributes: [SublimeUIAttributes.selected]
          }
        ],
        fg: findEditorColorForField(['list.activeSelectionForeground'])
      }
    ]
  }
}
