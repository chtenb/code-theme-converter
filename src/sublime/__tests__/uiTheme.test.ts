import { CodeTheme } from 'src/util/vscode'
import { createAdaptiveTheme } from '../uiTheme'
import mockTheme from './__fixtures__/test-theme.json'

describe('uiTheme', () => {
  describe('createAdaptiveTheme', () => {
    it('should be defined', () => {
      expect(createAdaptiveTheme(mockTheme as CodeTheme)).toBeDefined()
    })
    it('should create an adaptive theme', () => {
      expect(createAdaptiveTheme(mockTheme as CodeTheme)).toMatchSnapshot()
    })
  })
})
