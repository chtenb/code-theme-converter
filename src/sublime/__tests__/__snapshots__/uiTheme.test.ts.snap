// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uiTheme createAdaptiveTheme should create an adaptive theme 1`] = `
Object {
  "extends": "Adaptive.sublime-theme",
  "rules": Array [
    Object {
      "class": "title_bar",
      "fg": "#9da5b4",
    },
    Object {
      "class": "sidebar_container",
      "layer0.tint": "#0c0e13",
    },
    Object {
      "class": "status_bar",
      "layer0.tint": "#0c0e13",
    },
    Object {
      "class": "label_control",
      "fg": "#64727f",
      "parents": Array [
        Object {
          "class": "status_bar",
        },
      ],
    },
    Object {
      "attributes": Array [
        "selectable",
        "hover",
      ],
      "class": "tree_row",
      "layer0.opacity": 1,
      "layer0.tint": "#67769635",
    },
    Object {
      "attributes": Array [
        "selected",
      ],
      "class": "tree_row",
      "layer0.opacity": 1,
      "layer0.tint": "#67769640",
    },
    Object {
      "class": "sidebar_heading",
      "fg": "#c0c5cd",
    },
    Object {
      "class": "sidebar_label",
      "fg": "#7E8C99",
    },
    Object {
      "class": "sidebar_label",
      "fg": "#dfe1e8",
      "parents": Array [
        Object {
          "attributes": Array [
            "selected",
          ],
          "class": "tree_row",
        },
      ],
    },
    Object {
      "class": "vcs_status_badge",
      "layer0.tint": "#8892a0",
      "parents": Array [
        Object {
          "attributes": Array [
            "untracked",
          ],
          "class": "file_system_entry",
        },
      ],
    },
    Object {
      "class": "vcs_status_badge",
      "layer0.tint": "#7a9bc2",
      "parents": Array [
        Object {
          "attributes": Array [
            "modified",
          ],
          "class": "file_system_entry",
        },
      ],
    },
    Object {
      "class": "vcs_status_badge",
      "layer0.tint": "#516a88",
      "parents": Array [
        Object {
          "attributes": Array [
            "missing",
          ],
          "class": "file_system_entry",
        },
      ],
    },
    Object {
      "class": "vcs_status_badge",
      "layer0.tint": "#267fb5",
      "parents": Array [
        Object {
          "attributes": Array [
            "staged",
          ],
          "class": "file_system_entry",
        },
      ],
    },
    Object {
      "class": "vcs_status_badge",
      "layer0.tint": "#267fb5",
      "parents": Array [
        Object {
          "attributes": Array [
            "added",
          ],
          "class": "file_system_entry",
        },
      ],
    },
    Object {
      "class": "vcs_status_badge",
      "layer0.tint": "#516a88",
      "parents": Array [
        Object {
          "attributes": Array [
            "deleted",
          ],
          "class": "file_system_entry",
        },
      ],
    },
    Object {
      "class": "sidebar_label",
      "color": "#64727f",
      "parents": Array [
        Object {
          "attributes": Array [
            "ignored",
          ],
          "class": "file_system_entry",
        },
      ],
    },
    Object {
      "attributes": Array [
        "selected",
      ],
      "class": "tab_control",
      "tint_modifier": "#14191f",
    },
    Object {
      "attributes": Array [
        "dirty",
      ],
      "class": "tab_control",
      "layer2.tint": "#267fb5",
      "settings": Array [
        "highlight_modified_tabs",
      ],
    },
    Object {
      "class": "tab_label",
      "fg": "#7E8C99",
    },
    Object {
      "class": "tab_label",
      "fg": "#dfe1e8",
      "parents": Array [
        Object {
          "attributes": Array [
            "selected",
          ],
          "class": "tab_control",
        },
      ],
    },
  ],
  "variables": Object {
    "font_face": "system",
    "font_size": 12,
    "font_size_lg": 13,
    "font_size_sm": 11,
  },
}
`;
